File: app/backend/api/src/main/java/com/example/sms/domain/model/master/partner/Partner.java
package com.example.sms.domain.model.master.partner;

import com.example.sms.domain.model.common.address.Address;
import com.example.sms.domain.model.master.partner.customer.Customer;
import com.example.sms.domain.model.master.partner.vendor.Vendor;
import com.example.sms.domain.type.partner.MiscellaneousType;
import com.example.sms.domain.type.partner.TradeProhibitedFlag;
import com.example.sms.domain.type.partner.VendorType;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import lombok.Value;

import java.util.Collections;
import java.util.List;

/**
 * 取引先
 */
@Value
@AllArgsConstructor
@NoArgsConstructor(force = true)
public class Partner {
    PartnerCode partnerCode;          // 取引先コード
    PartnerName partnerName;          // 取引先名
    VendorType vendorType;        // 仕入先区分
    Address address;             // 住所
    TradeProhibitedFlag tradeProhibitedFlag; // 取引禁止フラグ
    MiscellaneousType miscellaneousType;   // 雑区分
    PartnerGroupCode partnerGroupCode;     // 取引先グループコード
    Credit credit;               // 与信
    List<Customer> customers;   // 顧客
    List<Vendor> vendors;       // 仕入先

    public static Partner of(
            String partnerCode,
            String partnerName,
            String partnerNameKana,
            Integer vendorType,
            String postalCode,
            String prefecture,
            String address1,
            String address2,
            Integer tradeProhibitedFlag,
            Integer miscellaneousType,
            String partnerGroupCode,
            Integer creditLimit,
            Integer temporaryCreditIncrease
    ) {
        return new Partner(
                PartnerCode.of(partnerCode),
                PartnerName.of(partnerName, partnerNameKana),
                VendorType.fromCode(vendorType),
                Address.of(postalCode, prefecture, address1, address2),
                TradeProhibitedFlag.fromCode(tradeProhibitedFlag),
                MiscellaneousType.fromCode(miscellaneousType),
                PartnerGroupCode.of(partnerGroupCode),
                Credit.of(creditLimit, temporaryCreditIncrease),
                List.of(),
                List.of()
        );
    }

    public static Partner of(
            PartnerCode partnerCode,
            PartnerName partnerName,
            VendorType vendorType,
            Address address,
            TradeProhibitedFlag tradeProhibitedFlag,
            MiscellaneousType miscellaneousType,
            PartnerGroupCode partnerGroupCode,
            Credit credit,
            List<Customer> customers,
            List<Vendor> vendors
    ) {
        return new Partner(
                partnerCode,
                partnerName,
                vendorType,
                address,
                tradeProhibitedFlag,
                miscellaneousType,
                partnerGroupCode,
                credit,
                customers,
                vendors
        );
    }

    public static Partner ofWithCustomers(Partner partner, List<Customer> customers) {
        return new Partner(
                partner.partnerCode,
                partner.partnerName,
                partner.vendorType,
                partner.address,
                partner.tradeProhibitedFlag,
                partner.miscellaneousType,
                partner.partnerGroupCode,
                partner.credit,
                customers,
                Collections.emptyList()
        );
    }

    public static Partner ofWithVendors(Partner partner, List<Vendor> vendors) {
        return new Partner(
                partner.partnerCode,
                partner.partnerName,
                partner.vendorType,
                partner.address,
                partner.tradeProhibitedFlag,
                partner.miscellaneousType,
                partner.partnerGroupCode,
                partner.credit,
                Collections.emptyList(),
                vendors
        );
    }
}

File: app/frontend/src/models/master/partner/partner.ts
import {PageNationType} from "../../../views/application/PageNation.tsx";
import {CustomerType} from "./customer.ts";
import {VendorType} from "./vendor.ts";
import {PrefectureEnumType} from "../shared.ts";

export type PartnerType = {
    partnerCode: string; // 取引先コード
    partnerName: string; // 取引先名
    partnerNameKana: string; // 取引先名カナ
    vendorType: VendorEnumType; // 仕入先区分
    postalCode: string; // 郵便番号
    prefecture: PrefectureEnumType; // 都道府県
    address1: string; // 住所1
    address2: string; // 住所2
    tradeProhibitedFlag: TradeProhibitedFlagEnumType; // 取引禁止フラグ
    miscellaneousType: MiscellaneousEnumType; // 雑区分
    partnerGroupCode: string; // 取引先グループコード
    creditLimit: number; // 与信限度額
    temporaryCreditIncrease: number; // 与信一時増加枠
    customers: CustomerType[]; // 取引先顧客リスト
    vendors: VendorType[]; // 取引先仕入先リスト
    checked: boolean;
};


export enum VendorEnumType {
    仕入先でない = "仕入先でない",
    仕入先 = "仕入先",
}

export enum TradeProhibitedFlagEnumType {
    OFF = "OFF", ON = "ON"
}

export enum MiscellaneousEnumType {
    対象外 = "対象外", 対象 = "対象"
}

export type PartnerCriteriaType = {
    partnerCode?: string; // 取引先コード
    partnerName?: string; // 取引先名
    partnerNameKana?: string; // 取引先名カナ
    vendorType?: number; // 仕入先区分 (数値)
    postalCode?: string; // 郵便番号
    prefecture?: string; // 都道府県
    address1?: string; // 住所1
    address2?: string; // 住所2
    tradeProhibitedFlag?: string; // 取引禁止フラグ (数値)
    miscellaneousType?: string; // 雑区分 (数値)
    partnerGroupCode?: string; // 取引先グループコード
    creditLimit?: string; // 与信限度額 (数値)
    temporaryCreditIncrease?: number; // 与信一時増加枠 (数値)
};

export type PartnerFetchType = {
    list: PartnerType[]; // 取引先リスト
} & PageNationType;

export const mapToPartnerResource = (partner: PartnerType): PartnerType => {
    return {
        ...partner,
    }
}

export const mapToPartnerCriteriaResource = (partnerCriteria: PartnerCriteriaType): PartnerCriteriaType => {
    const isEmpty = (value: unknown) => value === "" || value === null || value === undefined;
    return {
        ...(isEmpty(partnerCriteria.partnerCode) ? {} : { partnerCode: partnerCriteria.partnerCode }),
        ...(isEmpty(partnerCriteria.partnerName) ? {} : { partnerName: partnerCriteria.partnerName }),
        ...(isEmpty(partnerCriteria.partnerNameKana) ? {} : { partnerNameKana: partnerCriteria.partnerNameKana }),
        ...(isEmpty(partnerCriteria.vendorType) ? {} : { vendorType: partnerCriteria.vendorType }),
        ...(isEmpty(partnerCriteria.postalCode) ? {} : { postalCode: partnerCriteria.postalCode }),
        ...(isEmpty(partnerCriteria.prefecture) ? {} : { prefecture: partnerCriteria.prefecture }),
        ...(isEmpty(partnerCriteria.address1) ? {} : { address1: partnerCriteria.address1 }),
        ...(isEmpty(partnerCriteria.address2) ? {} : { address2: partnerCriteria.address2 }),
        ...(isEmpty(partnerCriteria.tradeProhibitedFlag) ? {} : { tradeProhibitedFlag: partnerCriteria.tradeProhibitedFlag }),
        ...(isEmpty(partnerCriteria.miscellaneousType) ? {} : { miscellaneousType: partnerCriteria.miscellaneousType }),
        ...(isEmpty(partnerCriteria.partnerGroupCode) ? {} : { partnerGroupCode: partnerCriteria.partnerGroupCode }),
        ...(isEmpty(partnerCriteria.creditLimit) ? {} : { creditLimit: partnerCriteria.creditLimit }),
        ...(isEmpty(partnerCriteria.temporaryCreditIncrease) ? {} : { temporaryCreditIncrease: partnerCriteria.temporaryCreditIncrease }),
    }
}
